deps/jemalloc/src/jemalloc.c

```diff
@@ -3927,5 +3927,145 @@ jemalloc_postfork_child(void) {
+}
+
+static long long ustime(void) {
+    struct timespec ts;
+    int res = clock_gettime(CLOCK_MONOTONIC, &ts);
+    assert(res == 0);
+    return (ts.tv_sec * 1000000) + (ts.tv_nsec / 1000);
+}
+
+static void init_defrag_bin_step(bin_t *bin, long long start_time, long long max_time) {
+    for (;bin->defrag_slabs_to_retain > 0; bin->defrag_slabs_to_retain--) {
+        extent_t *slab = extent_heap_remove_first(&bin->slabs_nonfull);
+        if (slab == NULL) {
+            bin->defrag_slabs_to_retain = 0;
+            break;
+        }
+
+        extent_heap_insert(&bin->slabs_nonfull_temp, slab);
+        if (!bin->highest_slab_to_retain_inited || extent_snad_comp(slab, &bin->highest_slab_to_retain) > 0) {
+            bin->highest_slab_to_retain = *slab;
+            bin->highest_slab_to_retain_inited = true;
+        }
+        if (ustime() - start_time >= max_time)
+            break;
+    }
+
+    if (bin->defrag_slabs_to_retain == 0) {
+        extent_t *r0 = extent_heap_first(&bin->slabs_nonfull);
+        extent_t *r1 = extent_heap_first(&bin->slabs_nonfull_temp);
+        phn_merge(extent_t, ph_link, r0, r1, extent_snad_comp, bin->slabs_nonfull.ph_root);
+        /* Clear the temp heap after merge */
+        extent_heap_new(&bin->slabs_nonfull_temp);
+        assert(extent_heap_empty(&bin->slabs_nonfull_temp));
+        bin->initing_defrag = false;
+    }
 }
+
+JEMALLOC_EXPORT void JEMALLOC_NOTHROW
+init_defrag(void) {
+    unsigned long long i;
+    tsd_t *tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+        const bin_info_t *bin_info = &bin_infos[i];
+        uint64_t used_slabs = bin->stats.nonfull_slabs + (bin->slabcur ? 1 : 0);
+        uint64_t full_slabs = bin->stats.curslabs - bin->stats.nonfull_slabs - (bin->slabcur ? 1 : 0);
+        uint64_t needed_slabs = (bin->stats.curregs - full_slabs * bin_info->nregs) / bin_info->nregs;
+        if ((bin->stats.curregs - full_slabs * bin_info->nregs) % bin_info->nregs) needed_slabs += 1;
+        needed_slabs += full_slabs;
+        used_slabs += full_slabs;
+        assert(bin->highest_slab_to_retain_inited == false);
+        assert(extent_heap_empty(&bin->slabs_nonfull_temp));
+        if (used_slabs != needed_slabs) {
+            int64_t nonfull_to_retain = needed_slabs - full_slabs;
+            if (bin->slabcur) nonfull_to_retain--;
+            bin->defrag_slabs_to_retain = nonfull_to_retain;
+            bin->initing_defrag = true;
+        } else {
+            bin->defrag_slabs_to_retain = 0;
+            bin->initing_defrag = false;
+        }
+        malloc_mutex_unlock(tsd_tsdn(tsd), &bin->lock);
+    }
+}
+
+JEMALLOC_EXPORT int JEMALLOC_NOTHROW
+init_defrag_step(long long max_time) {
+    unsigned long long i;
+    tsd_t * tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
```

Comment:
AFAIK, arena is thread-bind, since defrag is only run in the main thread, could this lead to overlooking internal fragmentation caused by threads?

Reply:
I think you mean external-fragmentation, but yes, you are right, defrag will only handle main thread allocations. But these are the Redis keys and values which are what we're aiming at. This isn't a change from the existing ative-defrag mechanism (@oranagra am I right here?).

Reply:
@yoav-steinberg i didn't look at the code of this PR yet, but i suspect you're wrong.
In redis both the IO threads and modules can current use allocations from other arenas, and that can actually be one of the causes for high fragmentation.
the original active-defrag mechanism handled these well, since for each allocation it used to compare the current slap frag ratio to the arena average bin frag as an indication for the need to re-allocate.

it's true that the re-allocate itself, would later move it to another arena, but i think that would have eventually reduce fragmentation.

If the new code only attempts to defrag allocations of the main arena, i think this is something we better fix.

Reply:
Some analysis:

I think our key/value allocations are from the main arena even when using I/O threads. This is because the query buffers aren't used for the key/values (even if they are used for argv).  

In case of modules, however, we can't assume this. So you are right there's a potential bug here. 

I was wrong re the old algorithm. I see it only looks at the slab's arena in order to make decisions and not at the main 
arena. But this means the old algorithm is less efficient in the case the allocation wasn't from the main arena. In that case we can simply return 1 from `get_defrag_hint()` since any fragmentation in the non-main arena will never resolve itself if we realloc to the main arena.

So bottom line:
* In the old algorithm we should always consider allocations in non-main arena's non-full stabs as hits.
* In new algorithm we should do the same.
* Ideally we should defrag each arena separately and then do the reallocations inside the same arena.

Reply:
> I think our key/value allocations are from the main arena even when using I/O threads. This is because the query buffers aren't used for the key/values (even if they are used for argv).

    We may create `c->argv` in the non-main thread, and execute the client command in the main thread, then
    we may just increment the `refcount` of robj in `c->argv` to create key or value, then this memory will be
    created by non-main thread.
    Please point out if I'm wrong

Reply:
Yoav was referring to the fact that key names and value of most types are `sds`, so they're not the same `robj` allocation from the argv. and for strings we call `tryObjectEncoding`, which i might told him that copies the object, but i now realize that i might have confused what `trimStringObjectIfNeeded` does (it copies the object only if it has a lot of excess space).
but anyway the final conclusion is the same, since modules may RM_RetainString these argv items and put them in the key-space, and modules can participate in the defrag mechanism (they have an API for that). so either way, we need to defrag all arenas.

Reply:
@oranagra @sundb I updated to code to handle multiple arenas. I did some local testing and this seems to work fine. Please re-review.